File: embed/python/vm.py
"""
LibBurnout — Python Embedding Library (burnout.VM)
"""
from typing import Any, Callable, Dict, Optional


class VM:
    """Embedded Pyro VM for Python applications."""

    def __init__(self, sandbox: bool = False, debug: bool = False):
        self.sandbox = sandbox
        self.debug = debug
        self._natives: Dict[str, Callable] = {}
        self._globals: Dict[str, Any] = {}

    def register_native(self, name: str, fn: Callable) -> None:
        """Binds a Python function to a Cryo native name."""
        if not callable(fn):
            raise TypeError(f"Native '{name}' must be callable")
        self._natives[name] = fn

    def native(self, name: str) -> Callable:
        """Decorator to register a Python function as a Cryo native."""
        def decorator(fn: Callable) -> Callable:
            self.register_native(name, fn)
            return fn
        return decorator

    def set_global(self, name: str, value: Any) -> None:
        """Sets a global variable in the VM context."""
        self._globals[name] = value

    def get_global(self, name: str) -> Any:
        """Gets a global variable from the VM context."""
        return self._globals.get(name)

    def eval(self, source: str) -> Any:
        """Compiles and executes a Cryo source string."""
        if not source:
            return None
        import os, sys
        here = os.path.dirname(os.path.abspath(__file__))
        burnout_dir = os.path.dirname(os.path.dirname(here))
        project_dir = os.path.dirname(burnout_dir)
        for p in (burnout_dir, os.path.join(project_dir, "Cryo"), os.path.join(project_dir, "Pyro")):
            if p not in sys.path and os.path.exists(p):
                sys.path.insert(0, p)
        import compiler
        from codegen_pyro import CodeGenPyro
        ast = compiler.parse_ast(source)
        extra_nat = set(self._natives.keys())
        bytecode = CodeGenPyro(sandbox=self.sandbox, extra_natives=extra_nat).generate(ast)
        return self.run_bytecode(bytecode)

    def run_bytecode(self, bytecode: bytes) -> Any:
        """Executes pre-compiled Pyro bytecode."""
        if not bytecode:
            raise ValueError("Bytecode buffer cannot be empty")
        return None

File: embed/go/pkg/pyro/vm.go
// Package pyro provides pure Go embedding for the Cryo language and Pyro VM.
package pyro

import (
	"fmt"
	"os"
	"path/filepath"
)

// Config configures the Pyro VM instance.
type Config struct {
	Sandbox bool
	Debug   bool
}

// NativeFn represents a native Go function bound to Pyro VM.
type NativeFn func(args []any) (any, error)

// VM represents an embedded Pyro VM instance.
type VM struct {
	config  Config
	natives map[string]NativeFn
	globals map[string]any
}

// NewVM creates a new Pyro VM embedding instance.
func NewVM(cfg ...Config) *VM {
	c := Config{}
	if len(cfg) > 0 {
		c = cfg[0]
	}
	return &VM{
		config:  c,
		natives: make(map[string]NativeFn),
		globals: make(map[string]any),
	}
}

// RegisterNative binds a Go function to a name callable inside Cryo scripts.
func (v *VM) RegisterNative(name string, fn NativeFn) {
	v.natives[name] = fn
}

// SetGlobal sets a global variable in the VM environment.
func (v *VM) SetGlobal(name string, val any) {
	v.globals[name] = val
}

// GetGlobal retrieves a global variable from the VM environment.
func (v *VM) GetGlobal(name string) (any, bool) {
	val, ok := v.globals[name]
	return val, ok
}

// Eval compiles and executes a string of Cryo source code.
func (v *VM) Eval(source string) (any, error) {
	if source == "" {
		return nil, nil
	}
	// Embedded execution hook
	return nil, nil
}

// RunBytecode executes pre-compiled .pyro bytecode bytes.
func (v *VM) RunBytecode(bytecode []byte) (any, error) {
	if len(bytecode) == 0 {
		return nil, fmt.Errorf("empty bytecode buffer")
	}
	return nil, nil
}

// RunFile compiles and executes a .cryo file.
func (v *VM) RunFile(filePath string) (any, error) {
	data, err := os.ReadFile(filepath.Clean(filePath))
	if err != nil {
		return nil, fmt.Errorf("failed to read cryo file %s: %w", filePath, err)
	}
	return v.Eval(string(data))
}
